p2p/peerinfo/peerinfo.go

```diff
@@ -0,0 +1,389 @@
+	// [2] is prev value for rate 2
+	bytesSent     [3]int64
+	bytesReceived [3]int64
+	// [0] is for the shorter-interval rate, [1] is for the longer-interval rate
+	recvRate [2]int64
+	sendRate [2]int64
+}
+
+func rateIndex(which int) int {
+	switch which {
+	case 1, 2:
+		return which - 1
+	default:
+		panic("bad rate index")
+	}
+}
+
+func bpsInterval(which int) time.Duration {
+	switch which {
+	case 1:
+		return bpsInterval1
+	case 2:
+		return bpsInterval2
+	default:
+		panic("bad rate index")
+	}
+}
+
+func (ds *DataStats) Tick(which int) {
+	ds.Lock()
+	defer ds.Unlock()
+	interval := bpsInterval(which)
+	idx := rateIndex(which)
+	ds.recvRate[idx] = int64(float64(ds.bytesReceived[0]-ds.bytesReceived[which]) / interval.Seconds())
+	ds.sendRate[idx] = int64(float64(ds.bytesSent[0]-ds.bytesSent[which]) / interval.Seconds())
+	ds.bytesSent[which] = ds.bytesSent[0]
+	ds.bytesReceived[which] = ds.bytesReceived[0]
+}
+
+func (ds *DataStats) RecvRate(which int) int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.recvRate[rateIndex(which)]
+}
+
+func (ds *DataStats) SendRate(which int) int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.sendRate[rateIndex(which)]
+}
+
+func (ds *DataStats) RecordSent(n int64) {
+	ds.Lock()
+	defer ds.Unlock()
+	ds.bytesSent[0] += n
+}
+
+func (ds *DataStats) RecordReceived(n int64) {
+	ds.Lock()
+	defer ds.Unlock()
+	ds.bytesReceived[0] += n
+}
+
+func (ds *DataStats) BytesSent() int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.bytesSent[0]
+}
+
+func (ds *DataStats) BytesReceived() int64 {
+	ds.Lock()
+	defer ds.Unlock()
+	return ds.bytesReceived[0]
+}
+
+type Info struct {
+	DataStats
+	connKinds   sync.Map
+	ClientStats PeerRequestStats
+	ServerStats PeerRequestStats
+}
+
+func (i *Info) Kind(c network.Conn) Kind {
+	k, ok := i.connKinds.Load(c.ID())
+	if !ok {
+		return KindUknown
+	}
+	return k.(Kind)
+}
+
+func (i *Info) SetKind(c network.Conn, k Kind) {
+	i.connKinds.Store(c.ID(), k)
+}
+
+// PeerInfo provides peer-related connection status and statistics.
+type PeerInfo interface {
+	// EnsurePeerInfo returns Info structure for the specific peers.
+	// If there's no such structure assigned for the peer yet, it creates one. The
+	// Info structure will be removed when all connections to the peer are closed.
+	EnsurePeerInfo(p peer.ID) *Info
+	// RecordReceived records that n bytes of data has been received from peer p
+	// via protocol proto.
+	RecordReceived(n int64, proto protocol.ID, p peer.ID)
+	// RecordReceived records that n bytes of data has been sent to the peer p
+	// via protocol proto.
+	RecordSent(n int64, proto protocol.ID, p peer.ID)
+	// Protocols returns the list of protocols used so far, in no particular order.
+	Protocols() []protocol.ID
+	// EnsureProtoStats returns DataStats structure for the specified protocol,
+	// allocating one if it doesn't exist yet.
+	EnsureProtoStats(proto protocol.ID) *DataStats
+}
+
+type PeerInfoTracker struct {
+	sync.Mutex
```

Comment:
same here, I don't think `Lock` and `Unlock` should be publicly accessible.

Reply:
Done